src/backend/executor/nodeDML.c

```diff
@@ -281,17 +281,29 @@ ExecInitDML(DML *node, EState *estate, int eflags)
 			dmlstate->cleanedUpSlot);
 
 	/*
-	 * We don't maintain typmod in the targetlist, so we should fixup the
-	 * junkfilter to use the same tuple descriptor as the result relation.
-	 * Otherwise the mismatch of tuple descriptor will cause a break in
-	 * ExecInsert()->reconstructMatchingTupleSlot().
+	 * The comment below is related to ExecInsert() and ExecUpdate(). The code
+	 * works correctly, because insert and update operations always translate
+	 * full set of attrs to targetlist. So, resulting tupledesc has the same
+	 * number of attrs after replacing. ExecDelete() doesn't have such
+	 * machinery, and more, can work with subset of table attrs. So, the code
+	 * below can add attrs, which doesn't exist in original targetlist. This
+	 * may cause execution errors.
 	 */
-	TupleDesc cleanTupType = CreateTupleDescCopy(dmlstate->ps.state->es_result_relation_info->ri_RelationDesc->rd_att);
+	if (estate->es_plannedstmt->commandType != CMD_DELETE)
```

Comment:
We don't take into account the AFTER trigger?

Reply:
In my mind, and this is described by a comment, the following "hack" is about reconstructMatchingTupleSlot() used by INSERT and UPDATE only, not by DELETE. So it doesn't matter how much attrs DELETE's targetlist contains here.